src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/ProjectExtensions.cs

```diff
@@ -58,7 +59,110 @@ public static async ValueTask<bool> AppliesToProjectAsync(this Type type, IProje
                 }
             }
 
+            // Check whether the type has an [DiagnosticAnalyzer] attribute
+            // If one exists, the type only applies to the project if the language matches
+            if (!AppliesToLanguage(type, project))
+            {
+                return false;
+            }
+
             return true;
         }
+
+        /// <summary>
+        /// Checks to see if the type should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature.</param>
+        /// <param name="project">the project currently being upgraded.</param>
+        /// <returns>true if the object is described as matching the project's language. Objects not defning the language they support will be true as default.</returns>
+        private static bool AppliesToLanguage(Type type, IProject project)
+        {
+            // check to see if this is a code fixer
+            if (DoesCodeFixerFilterThisFeature(type, project))
+            {
+                return false;
+            }
+
+            // check to see if this is a configuration step or other feature using the ApplicableLanguageAttribute
+            if (DoesApplicableLanguageFilterThisFeature(type, project))
+            {
+                return false;
+            }
+
+            // if the attribute is not applied then this type defaults to opt into language support
+            return true;
+        }
+
+        /// <summary>
+        /// Checks to see if the type is a codefixer that should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature decorated with <see cref="ExportCodeFixProviderAttribute"/>.</param>
```

Comment:
Nit-pick: Other doc comments that we have capitalize the first letter of params, returns, etc. sentence style. Also, what's decorated in this case is a CodeFixProvider, not an upgrade-assistant specific feature.